halo2_proofs/src/plonk/vanishing/prover.rs

```diff
@@ -47,10 +52,30 @@ impl<C: CurveAffine> Argument<C> {
         transcript: &mut T,
     ) -> Result<Committed<C>, Error> {
         // Sample a random polynomial of degree n - 1
-        let mut random_poly = domain.empty_coeff();
-        for coeff in random_poly.iter_mut() {
-            *coeff = C::Scalar::random(&mut rng);
-        }
+        let n_threads = current_num_threads();
+        let needed_scalars = (1usize << domain.k() as usize) / n_threads;
```

Comment:
I believe this implementation will be incorrect whenever `n_threads` is not a factor of `n`.
Let's say `k = 10` and `n_threads = 6`.
We have `1 << 10 = 1024`, `1024 / 6 = 170` (note that there's a rounding happening here)
Then we have 6 chunks of 170, so the final vector has length `6 * 170 = 1020`.

This could be the reason why there's a test failure in the MacOS, as it may be using a different number of threads than the ubuntu github worker.

Reply:
This is a proposal that should fix this issue.  I have only made sure that it compiles but I haven't tested it.
```diff
--- a/halo2_proofs/src/plonk/vanishing/prover.rs
+++ b/halo2_proofs/src/plonk/vanishing/prover.rs
@@ -4,10 +4,7 @@ use ff::{Field, PrimeField};
 use group::Curve;
 use rand_chacha::ChaCha20Rng;
 use rand_core::{RngCore, SeedableRng};
-use rayon::{
-    current_num_threads,
-    prelude::{IntoParallelRefIterator, IntoParallelRefMutIterator, ParallelIterator},
-};
+use rayon::{current_num_threads, prelude::*};
 
 use super::Argument;
 use crate::{
@@ -53,9 +50,11 @@ impl<C: CurveAffine> Argument<C> {
     ) -> Result<Committed<C>, Error> {
         // Sample a random polynomial of degree n - 1
         let n_threads = current_num_threads();
-        let needed_scalars = (1usize << domain.k() as usize) / n_threads;
+        let n = 1usize << domain.k() as usize;
+        let n_chunks = n_threads + if n % n_threads != 0 { 1 } else { 0 };
+        let mut rand_vec = vec![C::Scalar::zero(); n];
 
-        let mut thread_seeds: Vec<ChaCha20Rng> = (0..n_threads)
+        let mut thread_seeds: Vec<ChaCha20Rng> = (0..n_chunks)
             .into_iter()
             .map(|_| {
                 let mut seed = [0u8; 32];
@@ -64,15 +63,14 @@ impl<C: CurveAffine> Argument<C> {
             })
             .collect();
 
-        let rand_vec: Vec<C::Scalar> = thread_seeds
+        thread_seeds
             .par_iter_mut()
-            .flat_map(|mut rng| {
-                (0..needed_scalars)
-                    .into_iter()
-                    .map(|_| C::Scalar::random(&mut rng))
-                    .collect::<Vec<C::Scalar>>()
-            })
-            .collect();
+            .zip_eq(rand_vec.par_chunks_mut(n / n_threads))
+            .for_each(|(mut rng, chunk)| {
+                chunk
+                    .iter_mut()
+                    .for_each(|v| *v = C::Scalar::random(&mut rng))
+            });
 
         let random_poly: Polynomial<C::Scalar, Coeff> = Polynomial::from_evals(rand_vec);
```

Reply:
Thanks so much for catching this Edu!

To sum up. I benched this impl and my fix using trim and computing always one extra limb of random scalars. And seems that this is getting a consistent ~30% perf regression.
I assume it might be due to the divisions, because I can't explain how this is possible otherwise...

Here what I did:
```rust
// Sample a random polynomial of degree n - 1
        let n_threads = current_num_threads();
        let needed_scalars = (1usize << domain.k() as usize) / n_threads;

        // We always round up the division by adding 1 extra seed.
        let mut thread_seeds: Vec<ChaCha20Rng> = (0..n_threads + 1)
            .into_iter()
            .map(|_| {
                let mut seed = [0u8; 32];
                rng.fill_bytes(&mut seed);
                ChaCha20Rng::from_seed(seed)
            })
            .collect();

        let mut rand_vec: Vec<C::Scalar> = thread_seeds
            .par_iter_mut()
            .flat_map(|mut rng| {
                (0..needed_scalars)
                    .into_iter()
                    .map(|_| C::Scalar::random(&mut rng))
                    .collect::<Vec<C::Scalar>>()
            })
            .collect::<Vec<C::Scalar>>();

        rand_vec.truncate(1usize << domain.k() as usize);

        let random_poly: Polynomial<C::Scalar, Coeff> = domain.coeff_from_vec(rand_vec);
```